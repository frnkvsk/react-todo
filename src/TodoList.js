/**
 TodoList - this component should render the NewTodoForm component and should render the list of Todo components. Place your state that contains all of the todos in this component.
 */
import React, {useState, useEffect} from 'react';
import List from '@material-ui/core/List';
import Grid from '@material-ui/core/Grid';
import FormDialog from './EditTodo';
import NewTodoForm from './NewTodoForm';
import Todo from "./Todo";
import { v4 as uuid } from 'uuid';
import { makeStyles } from '@material-ui/core/styles';


const useStyles = makeStyles((theme) => ({
  root: {
    width: '100%',
    fontSize: '28px',
  },
  todoForm: {
    width: '100%',
    backgroundColor: '#afc2cb',
  },
  todo: {
    width: '100%',
    backgroundColor: '#e3f2fd',    
  },
}));

const TodoList = () => {
  
  let [todos, setTodos] = useState(
    JSON.parse(localStorage.getItem('todos')) || {}  
  );
  useEffect(() => {
    localStorage.setItem('todos', JSON.stringify(todos));
  }, [todos]);

  let [editDialogVisible, setEditDialogVisible] = useState(false);
  let [editDialogId, setEditDialogId] = useState("");
  let [editDialogText, setEditDialogText] = useState("");

  const addNewTodo = newTodo => {
    todos[uuid()] = {todo: newTodo, completed: false};
    setTodos(Object.assign({}, todos));
  }
  const saveTodo = (todoId, newTodo) => {
    todos[todoId] = {todo: newTodo, completed: false};
    setTodos(Object.assign({}, todos));
    setEditDialogVisible(false);
  }
  const completedTodo = (id) => {
    todos[id].completed = !todos[id].completed;
    setTodos(Object.assign({}, todos));
  }

  const hideTodo = id => {
    delete todos[id];
    setTodos(Object.assign({}, todos));
  }

  const openEditTodo = (id) => {
    setEditDialogText(todos[id].todo);
    setEditDialogId(id);
    setEditDialogVisible(true);
  }  

  const classes = useStyles();

  return (
    <Grid className={classes.root} container direction="column" >
      <Grid item container>
        <Grid item xs={"auto"} md={2} />
        <Grid item xs={12} md={8} >
          <NewTodoForm addNewTodo={addNewTodo} />

          { editDialogVisible && 
              <FormDialog 
                open={editDialogVisible} 
                todoId={editDialogId} 
                todoText={editDialogText} 
                saveTodo={saveTodo}  /> 
          }
          
          <List className={classes.todoText}>
            { 
              Object.entries(todos).map(e => ( 
                <Todo 
                  id={e[0]}
                  key={e[0]} 
                  todoText={e[1].todo}
                  completed={e[1].completed}
                  completedTodo={completedTodo}
                  openEditTodo={openEditTodo}
                  hideTodo={hideTodo} />
              ))
            } 
          </List>
        </Grid>      
        <Grid item xs={"auto"} md={2} />
      </Grid>      
    </Grid>
  )

}

export default TodoList;