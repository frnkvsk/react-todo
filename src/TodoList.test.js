import React from 'react';
import { render, screen } from '@testing-library/react';
import TodoList from './TodoList';

// smoke test
it("Control renders without crashing", () => {
  render(<TodoList />);
});

// snapshot test
it("matches snapshot", () => {
  const {asFragment} = render(<TodoList />);
  expect(asFragment()).toMatchSnapshot();
});

it("has NewTodoForm", async () => {

  const { getByLabelText, getByText } = render(<TodoList />);
  
  expect(screen.getByLabelText('Add a new todo')).toBeInTheDocument();
  
  expect(screen.queryByText(/ADD TODO/i)).toBeInTheDocument();  
  
});