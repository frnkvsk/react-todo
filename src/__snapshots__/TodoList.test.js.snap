// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root makeStyles-root-10 MuiGrid-container MuiGrid-direction-xs-column"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-item"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-auto MuiGrid-grid-md-2"
      />
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-8"
      >
        <form
          autocomplete="off"
          class="makeStyles-container-13"
          novalidate=""
        >
          <div
            class="MuiFormControl-root MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
              data-shrink="false"
              for="form-text-color"
              id="form-text-color-label"
            >
              Add a new todo
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input"
                id="form-text-color"
                name="todo"
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root-15 MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="PrivateNotchedOutline-legendLabelled-17"
                >
                  <span>
                    Add a new todo
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button-14 MuiButton-containedPrimary"
            id="form-btn"
            tabindex="0"
            type="submit"
          >
            <span
              class="MuiButton-label"
            >
              Add Todo
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </form>
        <ul
          class="MuiList-root MuiList-padding"
        />
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-auto MuiGrid-grid-md-2"
      />
    </div>
  </div>
</DocumentFragment>
`;
